aten/src/ATen/native/cuda/StochasticRounding.cu

```diff
@@ -0,0 +1,53 @@
+#include <ATen/ATen.h>
+#include <ATen/native/cuda/stochastic_rounding.cuh>
+
+
+namespace at {
+namespace native {
+
+template <typename input_t, typename output_t>
+__global__ void stochastic_rounding_kernel(
+    const input_t* input,
+    output_t* output,
+    const int64_t numel,
+    std::pair<uint64_t, uint64_t> seed_and_offset) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  curandStatePhilox4_32_10_t state;
+  curand_init(seed_and_offset.first, tid, seed_and_offset.second, &state);
+
+  for (int64_t i = tid; i < numel; i += blockDim.x * gridDim.x) {
+    float inp = static_cast<float>(input[i]);
+    output[i] = round_stochastically<output_t>(inp, curand_uniform(&state));
+  }
+}
+
+Tensor stochastic_rounding_cuda(const Tensor& input, Generator gen_) {
+
+  Tensor output = at::empty_like(input, input.options().dtype(kHalf), input.suggest_memory_format());
```

Comment:
Do you need to check if `input` is contiguous? Because you are looping  on 0 to numel, and if the input has some broadcasting, then you will get out of boundary in your kernel.

Reply:
Yes, I do (I created this function mainly for the sake of unittest and I was not careful in that sense). thank you.

Reply:
Now checking the contiguity https://github.com/csarofeen/pytorch/pull/17/files#diff-feb2a89ccaa81c81b69bf4c91e8f60a3R26